### PR TITLE
Implement packet clearing in relayer next

### PR DESCRIPTION
**Pull request by ljoss17**
Closes: #XXX

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer-cosmos/src/methods/unreceived_packet.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        event_id: WithBlockDataType::SendPacket,
+        source_channel_id: channel_id.clone(),
+        source_port_id: port_id.clone(),
+        destination_channel_id: counterparty_channel_id.clone(),
+        destination_port_id: counterparty_port_id.clone(),
+        sequences: sequences.to_vec(),
+        height: Qualified::SmallerEqual(QueryHeight::Specific(*height)),
+    };
+    let mut events = vec![];
+    for sequence in sequences.iter() {
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer-cosmos/src/methods/unreceived_packet.rs:108]: From what I understand `query_send_packet_events` from `crates/relayer/src/link/packet_events.rs` does the following:

1. Build the `QueryPacketEventDataRequest` and call `query_packet_events`
2. If the height is `Qualified::Equal` query the packets from block (not currently done in the this implementation for relayer-next)
3. If the height is `Qualified::SmallerEqual` query the packets from TXs
4. `query_packet_from_txs` calls `tx_search(packet_query(..))` for each sequence number, similar to what is being done in this implementation
	1. With the addition of mapping the ABCI events to `IbcEventWithHeight` by calling `packet_from_tx_search_response`.

I don't see how this is more efficient, and the issue is that the cosmos implementation uses ABCI Events as its `Event` type so `try_extract_send_packet_event` expects an ABCI Event and not a `IbcEventWithHeight`.

And `query_packet_events_with` from `crates/relayer/src/link/packet_events.rs` is called with `query_send_packet_events` and `query_write_ack_events` as `query_fn`, so it also ends up calling `query_packet_events`.

File: crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    HasIbcChainTypes<Counterparty> + HasIbcPacketTypes<Counterparty> + HasErrorType
+where
+    Counterparty: HasIbcChainTypes<Self>,
+{
+    async fn query_unreceived_packet_sequences(
+        &self,
+        channel_id: &Self::ChannelId,
+        port_id: &Self::PortId,
+        sequences: &[Counterparty::Sequence],
+    ) -> Result<(Vec<Self::Sequence>, Self::Height), Self::Error>;
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs]: Fixed in https://github.com/informalsystems/hermes/pull/3488/commits/92c386ff67c5d0b0ba43b11c80d2055a4bd1cd74

File: crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        port_id: &Chain::PortId,
+        counterparty_channel_id: &Counterparty::ChannelId,
+        counterparty_port_id: &Counterparty::PortId,
+        sequences: &[Counterparty::Sequence],
+        height: &Counterparty::Height,
+    ) -> Result<Vec<Chain::OutgoingPacket>, Chain::Error>;
+}
+
+#[async_trait]
+pub trait CanQueryUnreceivedPackets<Counterparty>:
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs]: Fixed in https://github.com/informalsystems/hermes/pull/3488/commits/11da2e4f25c3da2027d05395969c7ccef913b422

File: crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+where
+    Counterparty: HasIbcChainTypes<Self>,
+{
+    async fn query_unreceived_packets(
+        &self,
+        channel_id: &Self::ChannelId,
+        port_id: &Self::PortId,
+        counterparty_channel_id: &Counterparty::ChannelId,
+        counterparty_port_id: &Counterparty::PortId,
+        sequences: &[Counterparty::Sequence],
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs]: Fixed in https://github.com/informalsystems/hermes/pull/3488/commits/92c386ff67c5d0b0ba43b11c80d2055a4bd1cd74

File: tools/integration-test/src/tests/next/packet_clear.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                .await
+                .unwrap();
+
+            assert_eq!(dst_commitments, vec!());
+
+            info!("Assert query unreceived packet sequences works as expected");
+
+            let (unreceived_packet_sequences, height): (Vec<Sequence>, Height) =
+                CanQueryUnreceivedPacketSequences::query_unreceived_packet_sequences(
+                    relay_context.relay_a_to_b().src_chain(),
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply tools/integration-test/src/tests/next/packet_clear.rs]: Fixed in https://github.com/informalsystems/hermes/pull/3488/commits/e0b57276ffac462d71f87fe62aa5d397711560d6

File: crates/relayer-components/src/chain/traits/queries/packet_commitments.rs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+use crate::core::traits::error::HasErrorType;
+use crate::std_prelude::*;
+
+#[async_trait]
+pub trait CanQueryPacketCommitments<Counterparty>:
+    HasIbcChainTypes<Counterparty> + HasErrorType
+where
+    Counterparty: HasIbcChainTypes<Self>,
+{
+    async fn query_packet_commitments(
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer-components/src/chain/traits/queries/packet_commitments.rs:17]: Fixed in https://github.com/informalsystems/hermes/pull/3488/commits/92c386ff67c5d0b0ba43b11c80d2055a4bd1cd74 and https://github.com/informalsystems/hermes/pull/3488/commits/132ae5d017275e9d40f921b77453a23ee0a622bd

File: crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        height: &Counterparty::Height,
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs]: Fixed in https://github.com/informalsystems/hermes/pull/3488/commits/495bffb8b2657adedecc163142a67552d091c7e7

File: crates/relayer-cosmos/src/methods/unreceived_packet.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (soareschen) [reply crates/relayer-cosmos/src/methods/unreceived_packet.rs:114]: Hmm you are right. Looks like the algorithm is ultimately based on `query_packets_from_txs` in `crates/relayer/src/chain/cosmos/query/tx.rs`. Maybe we should note that down and add a todo for investigating more efficient approach.

One more nit for this: it should be possible to use the [`Stream` construct with `.map()`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.map) to query for all events in parallel and then collect them into a vec.

[the remark this answers is not in the document]
Comment (soareschen) [reply crates/relayer-cosmos/src/methods/unreceived_packet.rs:114]: Probably a more structured way to define this is to first define a `CanQuerySendPacketsFromSequence` trait (without -s), and then implement `CanQuerySendPacketsFromSequences` by using `Stream::map` and calling the singular method for each sequence.

File: crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+#[async_trait]
+pub trait CanQuerySendPacketsFromSequences<Counterparty>:
+    HasIbcChainTypes<Counterparty> + HasIbcPacketTypes<Counterparty> + HasErrorType
+where
+    Counterparty: HasIbcChainTypes<Self>,
+{
+    /// Given a list of sequences, a channel and port will query a list of outgoing
+    /// packets which have not been relayed.
+    async fn query_unreceived_packets(
```
Comment (soareschen) [inline crates/relayer-components/src/chain/traits/queries/unreceived_packets.rs:55]: The name of the method should also change to `query_send_packets_from_sequences`. Also, let's move this trait to a separate file like `send_packets.rs`.

**Review by soareschen — CHANGES_REQUESTED** [review]

**Review by soareschen — APPROVED** [review]

**Pull request merged**